### PR TITLE
Add initial support for DataFlowSanitizer

### PR DESCRIPTION
**Pull request by rcvalle**
Adds initial support for DataFlowSanitizer to the Rust compiler. It currently supports `-Zsanitizer-dataflow-abilist`. Additional options for it can be passed to LLVM command line argument processor via LLVM arguments using `llvm-args` codegen option (e.g., `-Cllvm-args=-dfsan-combine-pointer-labels-on-load=false`).

### CODE REVIEW

**Review by browneee — COMMENTED** [review]
Hello,
(I am DFSan maintainer in LLVM.)

Thank you for working on DFSan support for Rust!

DFSan support will also need DFSan APIs so the program can make DFSan do something useful. Imo it would be best to do this in this commit, so we can properly test the DFSan pass is working correctly.

File: compiler/rustc_codegen_llvm/src/back/write.rs
Comment (rcvalle) [reply compiler/rustc_codegen_llvm/src/back/write.rs]: I couldn't combine them into one variable because it seems the array of pointers don't live long enough until its use, so I had to bind it to an additional variable. I changed it to `sanitize_dataflow_abilist_ptrs` as suggested.

File: compiler/rustc_codegen_ssa/src/back/write.rs
Comment (rcvalle) [reply compiler/rustc_codegen_ssa/src/back/write.rs:99]: This option currently uses `parse_list` which has a `Vec<String>` parameter. In this case, I could:

1. Leave as is.
2. Change it to `Vec<PathBuf>` and use `parse_opt_pathbuf` instead, but the option would change to the user being able to specify a single path only.
3. Create a `parse_pathbuf_list` similarly to `parse_list` that has a `Vec<PathBuf>` parameter.
4. Try to refactor `parse_list` to have a trait parameter instead so both `Vec<PathBuf>` and `Vec<String>` arguments could be passed (if possible).

Let me know what you think.

Comment (browneee) [reply compiler/rustc_codegen_ssa/src/back/write.rs:99]: Thanks for looking into it. Option 1 sounds ok to me. Probably more of a question for someone on the Rust side.

File: src/doc/unstable-book/src/compiler-flags/sanitizer.md
Comment (rcvalle) [reply src/doc/unstable-book/src/compiler-flags/sanitizer.md:33]: Done.

File: tests/codegen/sanitizer/dataflow-instrument-functions.rs
Comment (rcvalle) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: I don't think the Rust compiler currently has any FFI bindings for any of the sanitizers interfaces. Since those are user interfaces, I think the best way to provide them would be to create and provide a sanitizers crate with those bindings. What do you think?

Comment (browneee) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: Ideally these user interfaces would be kept in sync with the compiler.
I don't know how this should best fit together for Rust repos/build/release.

I guess simplest thing (external crate) is probably fine for getting it working.

Would probably be best to ask Rust people how it should fit together.

Comment (rcvalle) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: Another possibility is to have this crate in the `library/` directory along with `core`, `std`, and others. (Note that this crate is different than the `rustc_sanitizers` planned to be in the `compiler/` directory because those are user interfaces.) @cuviper @wesleywiser What do you think?

Comment (cuviper) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: I think it will make it a much bigger lift if we have to add to the standard library.

Can you sketch out what kind of calls the user code would need to make?

Comment (rcvalle) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: I put together a crate so you can see what it looks like: https://github.com/rcvalle/rust-sanitizers and https://crates.io/crates/sanitizers. The question is whether this crate should be moved into `library/` (not be part of `std` or `core`, but be placed alongside them similarly to `alloc``, `backtrace`, and others) or if it should stay as an external crate just as it was published now.

(The idea is to split this crate into sanitizers and sanitizers-sys in the future, and provide both the FFI bindings and an idiomatic Rust interface for the sanitizers interfaces.)

**Comment by nikic** [comment]
Do you have some specific use case in mind for this, or is this more for the sake of completeness? I'm always a bit fuzzy on what people actually use dfsan for.

File: tests/codegen/sanitizer/dataflow-instrument-functions.rs
Comment (nikic) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: For the purpose of *this* PR, I'd suggest to just add some minimal FFI bindings inside a test -- enough to demonstrate that it actually works, without blocking on the question of how to expose the API for now.

Prototyping this in an external crate first seems like a good idea in any case, regardless of whether it will be eventually included in the standard library.

File: compiler/rustc_codegen_ssa/src/back/write.rs
Comment (nikic) [inline compiler/rustc_codegen_ssa/src/back/write.rs:99]: Is the `Option<>` wrapper here necessary? (That is, does it encode something that is semantically different than an empty vec?)

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
Comment (nikic) [inline compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp:688]:
```suggestion
  char **SanitizeDataFlowABIList;
```

**Comment by browneee** [comment]
> Do you have some specific use case in mind for this, or is this more for the sake of completeness? I'm always a bit fuzzy on what people actually use dfsan for.

It helps us to do [C++ and Rust interop](https://security.googleblog.com/2024/02/improving-interoperability-between-rust-and-c.html). If a C++ codebase (that already uses DFSan) starts interoperating with Rust, the DFSan build will break - unless the Rust part can also support DFSan.

File: compiler/rustc_codegen_ssa/src/back/write.rs
Comment (rcvalle) [reply compiler/rustc_codegen_ssa/src/back/write.rs:99]: It doesn't do anything semantically different. The reason I used it is because there isn't a `parse_comma_list`, just a `parse_opt_comma_list`, but I could create one. What do you think?

File: tests/codegen/sanitizer/dataflow-instrument-functions.rs
Comment (rcvalle) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: Done.

File: compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp
Comment (rcvalle) [reply compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp]: Done.

File: compiler/rustc_codegen_ssa/src/back/write.rs
Comment (rcvalle) [reply compiler/rustc_codegen_ssa/src/back/write.rs]: I think it makes sense to add a `parse_comma_list` so I went ahead and added it.

Comment (rcvalle) [reply compiler/rustc_codegen_ssa/src/back/write.rs]: Done.

File: tests/ui/sanitize/sanitizer-dataflow-basic.rs
Comment (browneee) [inline tests/ui/sanitize/sanitizer-dataflow-basic.rs:39]: This could be a separate test...

I suggest send `i` on a trip through some Rust code to confirm the transformation works.

For example: pass it into a function, store it in a vec, add something to it, clone a reference of it, etc. and then confirm that the taint label was propagated through all that.

e.g.
```
fn propagate2(i: &i64) -> i64 {
  i.clone()
}
fn propagate(i: i64) -> i64 {
  let v = vec!(i, 1, 2, 3);
  let j = v.iter().sum();
  propagate2(&j)
}
```

```
  let mut x = propagate(i);
  let x_ptr = &mut x as *mut i64;
  let x_read_label = unsafe { dfsan_read_label(x_ptr as *const c_void, size_of::<i64>()) };
  assert_eq!(unsafe { dfsan_has_label(x_read_label, i_label) }, 1);
  assert_eq!(unsafe { dfsan_has_label(x_read_label, j_label) }, 1);
```

File: tests/ui/sanitize/sanitizer-dataflow-basic.rs
Comment (browneee) [inline tests/ui/sanitize/sanitizer-dataflow-basic.rs:6]: Can `-Zsanitizer=dataflow` be made to imply some default abilists?

So you would use `-Zsanitizer-dataflow-abilist=...` if you wanted something other than the default abilist files.

File: tests/ui/sanitize/sanitizer-dataflow-basic.rs
Comment (rcvalle) [reply tests/ui/sanitize/sanitizer-dataflow-basic.rs:49]: Done. I just replicated the basic functionality test from the DFSan tests in the LLVM Project for now. Do you think we should block this until we review and refactor this (and possibly other tests) or is it okay we do it later in a separate PR?

File: tests/ui/sanitize/sanitizer-dataflow-abilist-rust-no-core-no-std.txt
Comment (browneee) [inline tests/ui/sanitize/sanitizer-dataflow-abilist-rust-no-core-no-std.txt:12]: Do we need separate abilist files for no-core / no-std?

I would have thought core/std rust is a superset of no-core/no-std rust?

If so, I think we can have one ABI list for core/std configuration, then building no-core/no-std should just not hit any cases where we need to apply the abilist rules for symbols beyond no-core/no-std.

File: tests/ui/sanitize/sanitizer-dataflow-basic.rs
Comment (browneee) [reply tests/ui/sanitize/sanitizer-dataflow-basic.rs:49]: LGTM

File: tests/ui/sanitize/sanitizer-dataflow-basic.rs
Comment (rcvalle) [reply tests/ui/sanitize/sanitizer-dataflow-basic.rs:6]: I think for this we'll need to wait until the work for default ignorelists is done for all sanitizers and support for default ignorelists is added. What do you think?

File: tests/ui/sanitize/sanitizer-dataflow-abilist-rust-no-core-no-std.txt
Comment (rcvalle) [reply tests/ui/sanitize/sanitizer-dataflow-abilist-rust-no-core-no-std.txt:12]: Sorry, by no core and no std in the file name it means core and std are uninstrumented since we don't build tests with Cargo and thus can't use the Cargo build-std feature, not that these lists are for use with the `no_core` and `no_std` features. I also moved these ABI list files to test/sanitizer so it implies they should only beused by these tests. The actual ABI list files users should use are shipped with the 'sanitizers` crate.

Comment (rcvalle) [reply tests/ui/sanitize/sanitizer-dataflow-abilist-rust-no-core-no-std.txt]: I also removed them and consolidated the ABI list files into a single ABI list file for the tests without the no core and no std in the file name.

File: tests/ui/sanitize/sanitizer-dataflow-basic.rs
Comment (browneee) [reply tests/ui/sanitize/sanitizer-dataflow-basic.rs]: I think the default ignorelist is front-end specific. Clang already provides default abilists.

Or are you saying this can wait until someone does default lists for all rust sanitizers?

File: tests/ui/sanitize/sanitizer-dataflow-basic.rs
Comment (browneee) [reply tests/ui/sanitize/sanitizer-dataflow-basic.rs:49]: I think that can be later.

**Review by browneee — APPROVED** [review]

File: tests/ui/sanitize/sanitizer-dataflow-basic.rs
Comment (rcvalle) [reply tests/ui/sanitize/sanitizer-dataflow-basic.rs]: I've to add support for having default ignorelists to the frontend and add the default ignorelists for all sanitizers. I'm planning to do it later as part of our work in organizing support for sanitizers (see our [roadmap](https://hackmd.io/@rcvalle/S1Ou9K6H6)). I don't think this PR should block on this; otherwise, it'll take some time for it to be merged.

File: tests/codegen/sanitizer/dataflow-instrument-functions.rs
Comment (rcvalle) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: As pointed out to me by @cuviper and @wesleywiser, some of the GitHub workflows don't use the `src/llvm-project` LLVM, and since this test requires the new dfsan runtime library which is enabled by this PR to be built when the build sanitizers option is enabled for bootstrap, it will fail to link with the dfsan runtime library as it doesn't seem to be present in the LLVM used by the GitHub workflow.

This test passes locally  using the `src/llvm-project` LLVM and is also part of the `sanitizers` crate so I'll remove it from this PR.

**Review by browneee — APPROVED** [review]

File: src/tools/rustfmt/config_proc_macro/src/attrs.rs
Comment (nikic) [inline src/tools/rustfmt/config_proc_macro/src/attrs.rs:75]: Why did this file change?

File: tests/codegen/sanitizer/dataflow-instrument-functions.rs
Comment (rcvalle) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: Ahhh, I didn't know about that. Let me give it a try.

File: src/tools/rustfmt/config_proc_macro/src/attrs.rs
Comment (rcvalle) [reply src/tools/rustfmt/config_proc_macro/src/attrs.rs]: Apparently `./x.py fmt` caught this unformatted part and formatted it. I reverted these changes.

File: tests/codegen/sanitizer/dataflow-instrument-functions.rs
Comment (rcvalle) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: It worked! (It seems it also needed `needs-sanitizer-support` in the test file.) Thank you! I'll also adjust the path and file names to conform to #121811 which is in the queue to be merged. Then it should be good to go. I'll let you know! Thanks again!

Comment (rcvalle) [reply tests/codegen/sanitizer/dataflow-instrument-functions.rs:1]: #121811 was merged and I adjusted the path and file names. It should be good to go. Thanks again!

**Pull request merged**
